File: android/src/main/java/ai/founderos/mobiletracker/LocationManager.kt
package ai.founderos.mobiletracker

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority
import com.google.android.gms.tasks.CancellationTokenSource
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlin.coroutines.resume

/**
 * Location Manager for geolocation tracking
 * Web Reference: api.ts lines 300-362
 */
class LocationManager(
    private val context: Context,
    private val apiClient: ApiClient,
    private val config: TrackerConfig
) {
    private val fusedLocationClient: FusedLocationProviderClient =
        LocationServices.getFusedLocationProviderClient(context)
    
    /**
     * Request location update for a session
     * Web Reference: api.ts lines 300-322
     */
    suspend fun requestLocationUpdate(sessionId: String) = withContext(Dispatchers.Main) {
        // Check if location permissions are granted
        val fineLocationGranted = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
        
        val coarseLocationGranted = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_COARSE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
        
        if (!fineLocationGranted && !coarseLocationGranted) {
            if (config.debug) {
                println("[LocationManager] Location permissions not granted")
            }
            return@withContext
        }
        
        try {
            // Request current location
            val locationData = getCurrentLocation()
            
            if (locationData != null) {
                // Update session location on backend
                // Web Reference: api.ts line 314
                val success = apiClient.updateSessionLocation(sessionId, locationData)
                
                if (config.debug) {
                    if (success) {
                        println("[LocationManager] Session location updated successfully")
                    } else {
                        println("[LocationManager] Failed to update session location")
                    }
                }
            }
        } catch (e: Exception) {
            // Handle errors gracefully
            // Web Reference: api.ts lines 315-318
            if (config.debug) {
                println("[LocationManager] Error getting geolocation: ${e.message}")
            }
        }
    }
    
    /**
     * Get current location using FusedLocationProviderClient
     */
    @SuppressLint("MissingPermission")
    private suspend fun getCurrentLocation(): LocationData? = suspendCancellableCoroutine { continuation ->
        try {
            val cancellationTokenSource = CancellationTokenSource()
            
            // Set up cancellation
            continuation.invokeOnCancellation {
                cancellationTokenSource.cancel()
            }
            
            // Request current location with high accuracy
            fusedLocationClient.getCurrentLocation(
                Priority.PRIORITY_HIGH_ACCURACY,
                cancellationTokenSource.token
            ).addOnSuccessListener { location ->
                if (location != null) {
                    // Extract latitude, longitude, accuracy
                    // Web Reference: api.ts lines 313-316
                    val locationData = LocationData(
                        latitude = location.latitude,
                        longitude = location.longitude,
                        accuracy = location.accuracy.toDouble()
                    )
                    continuation.resume(locationData)
                } else {
                    if (config.debug) {
                        println("[LocationManager] Location is null")
                    }
                    continuation.resume(null)
                }
            }.addOnFailureListener { exception ->
                // Handle errors gracefully
                // Web Reference: api.ts lines 320-323
                if (config.debug) {
                    println("[LocationManager] Error getting geolocation: ${exception.message}")
                }
                continuation.resume(null)
            }
        } catch (e: SecurityException) {
            if (config.debug) {
                println("[LocationManager] Security exception: ${e.message}")
            }
            continuation.resume(null)
        }
    }
}

File: ios/MobileTracker/LocationManager.swift
import Foundation
import CoreLocation

/// Location Manager for geolocation tracking
/// Web Reference: api.ts lines 300-362
@available(iOS 13.0, macOS 10.15, *)
class LocationManager: NSObject, CLLocationManagerDelegate {
    private let locationManager: CLLocationManager
    private let apiClient: ApiClient
    private let config: TrackerConfig
    private var sessionId: String?
    private var locationContinuation: CheckedContinuation<LocationData?, Never>?
    
    /// Initialize LocationManager with API client and configuration
    init(apiClient: ApiClient, config: TrackerConfig) {
        self.apiClient = apiClient
        self.config = config
        self.locationManager = CLLocationManager()
        super.init()
        
        self.locationManager.delegate = self
        self.locationManager.desiredAccuracy = kCLLocationAccuracyBest
    }
    
    /// Request location update for a session
    /// Web Reference: api.ts lines 300-322
    func requestLocationUpdate(_ sessionId: String) async {
        self.sessionId = sessionId
        
        #if os(iOS)
        // Check if location services are available
        guard CLLocationManager.locationServicesEnabled() else {
            if config.debug {
                print("[LocationManager] Location services not enabled")
            }
            return
        }
        
        // Check authorization status
        let authStatus: CLAuthorizationStatus
        if #available(iOS 14.0, *) {
            authStatus = locationManager.authorizationStatus
        } else {
            authStatus = CLLocationManager.authorizationStatus()
        }
        
        switch authStatus {
        case .notDetermined:
            // Request permission when in use
            locationManager.requestWhenInUseAuthorization()
            return
        case .restricted, .denied:
            if config.debug {
                print("[LocationManager] Location access denied or restricted")
            }
            return
        case .authorizedWhenInUse, .authorizedAlways:
            // Permission granted, request location
            await requestCurrentLocation()
        @unknown default:
            if config.debug {
                print("[LocationManager] Unknown authorization status")
            }
            return
        }
        #else
        if config.debug {
            print("[LocationManager] Location services only available on iOS")
        }
        #endif
    }
    
    /// Request current location
    private func requestCurrentLocation() async {
        let locationData = await withCheckedContinuation { continuation in
            self.locationContinuation = continuation
            
            // Request a single location update
            if #available(iOS 14.0, *) {
                locationManager.requestLocation()
            } else {
                // For iOS 13, start updating location and stop after first update
                locationManager.startUpdatingLocation()
            }
        }
        
        // If we got location data, update the session
        if let locationData = locationData, let sessionId = self.sessionId {
            let success = await apiClient.updateSessionLocation(sessionId, location: locationData)
            
            if config.debug {
                if success {
                    print("[LocationManager] Session location updated successfully")
                } else {
                    print("[LocationManager] Failed to update session location")
                }
            }
        }
    }
    
    // MARK: - CLLocationManagerDelegate
    
    /// Handle location updates
    /// Web Reference: api.ts lines 312-318 (success callback)
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        // Stop updating location for iOS 13 compatibility
        if #available(iOS 14.0, *) {
            // iOS 14+ uses requestLocation() which stops automatically
        } else {
            manager.stopUpdatingLocation()
        }
        
        guard let location = locations.first else {
            locationContinuation?.resume(returning: nil)
            locationContinuation = nil
            return
        }
        
        // Extract latitude, longitude, accuracy
        // Web Reference: api.ts lines 313-316
        let locationData = LocationData(
            latitude: location.coordinate.latitude,
            longitude: location.coordinate.longitude,
            accuracy: location.horizontalAccuracy
        )
        
        locationContinuation?.resume(returning: locationData)
        locationContinuation = nil
    }
    
    /// Handle location errors
    /// Web Reference: api.ts lines 320-323 (error callback)
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        if config.debug {
            print("[LocationManager] Error getting geolocation: \(error.localizedDescription)")
        }
        
        locationContinuation?.resume(returning: nil)
        locationContinuation = nil
    }
    
    /// Handle authorization changes
    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        #if os(iOS)
        let authStatus: CLAuthorizationStatus
        if #available(iOS 14.0, *) {
            authStatus = manager.authorizationStatus
        } else {
            authStatus = CLLocationManager.authorizationStatus()
        }
        
        // If permission was just granted, request location
        if (authStatus == .authorizedWhenInUse || authStatus == .authorizedAlways),
           let sessionId = self.sessionId {
            Task {
                await requestLocationUpdate(sessionId)
            }
        }
        #endif
    }
}
